### PR TITLE
qgis: fix build of LTS release with proj>7

### PR DESCRIPTION
**Pull request by bernhardkaindl**
Apply upstream patch to qgis@3.16:3.19 to fix build with proj@8

### CODE REVIEW

**Comment by Sinan81** [comment]
@bernhardkaindl Could you please provide a reference for this PR as to why proj@8: isn't supported. I was guessing QGIS wouldn't support latest proj but due to build issues couldn't test it.

**Comment by Sinan81** [comment]
This seems relevant: https://github.com/qgis/QGIS/issues/41735

**Comment by bernhardkaindl** [comment]
> This seems relevant: [qgis/QGIS#41735](https://github.com/qgis/QGIS/issues/41735)

@Sinan81 added the upstream patch from it.

**Comment by Sinan81** [comment]
I think it would be more elegant to sidestep the shortcomings of FindProj.py and modify Cmake variables directly as follows within Spack:
```diff
diff --git a/var/spack/repos/builtin/packages/qgis/package.py b/var/spack/repos/builtin/packages/qgis/package.py
index ab7f60c..3fc57dd 100644
--- a/var/spack/repos/builtin/packages/qgis/package.py
+++ b/var/spack/repos/builtin/packages/qgis/package.py
@@ -161,7 +161,9 @@ def cmake_args(self):
                     self.spec['geos'].prefix.bin,
                     '-DGSL_CONFIG_PREFER_PATH=' + self.spec['gsl'].prefix.bin,
                     '-DPOSTGRES_CONFIG_PREFER_PATH=' +
-                    self.spec['postgresql'].prefix.bin
+                    self.spec['postgresql'].prefix.bin,
+                    '-DPROJ_INCLUDE_DIR=' + self.spec['proj'].prefix.include,
                     ])
 ```

**Comment by Sinan81** [comment]
Related to https://github.com/spack/spack/issues/26033#issuecomment-941872244

**Comment by bernhardkaindl** [comment]
> I think it would be more elegant to sidestep the shortcomings of FindProj.py and modify Cmake variables directly as follows within Spack:
> 
> ```diff
> diff --git a/var/spack/repos/builtin/packages/qgis/package.py b/var/spack/repos/builtin/packages/qgis/package.py
> index ab7f60c..3fc57dd 100644
> --- a/var/spack/repos/builtin/packages/qgis/package.py
> +++ b/var/spack/repos/builtin/packages/qgis/package.py
> @@ -161,7 +161,9 @@ def cmake_args(self):
>                      self.spec['geos'].prefix.bin,
>                      '-DGSL_CONFIG_PREFER_PATH=' + self.spec['gsl'].prefix.bin,
>                      '-DPOSTGRES_CONFIG_PREFER_PATH=' +
> -                    self.spec['postgresql'].prefix.bin
> +                    self.spec['postgresql'].prefix.bin,
> +                    '-DPROJ_INCLUDE_DIR=' + self.spec['proj'].prefix.include,
>                      ])
> ```

pushed.

File: var/spack/repos/builtin/packages/qgis/package.py [hunk 1]
```diff
@@ -161,7 +161,10 @@ def cmake_args(self):
                     self.spec['geos'].prefix.bin,
                     '-DGSL_CONFIG_PREFER_PATH=' + self.spec['gsl'].prefix.bin,
                     '-DPOSTGRES_CONFIG_PREFER_PATH=' +
-                    self.spec['postgresql'].prefix.bin
+                    self.spec['postgresql'].prefix.bin,
+                    '-DPROJ_INCLUDE_DIR=' + self.spec['proj'].prefix.include,
+                    '-DSQLITE3_LIBRARY='
+                    + join_path(self.spec['sqlite'].prefix.lib, 'libsqlite3.so')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/qgis/package.py]: This won't work if the library is in lib64 or you're on macOS, use this instead:
```suggestion
                    + self.spec['sqlite'].libs[0]
```

File: var/spack/repos/builtin/packages/qgis/package.py [hunk 2]
```diff
@@ -161,7 +161,10 @@ def cmake_args(self):
                     self.spec['geos'].prefix.bin,
                     '-DGSL_CONFIG_PREFER_PATH=' + self.spec['gsl'].prefix.bin,
                     '-DPOSTGRES_CONFIG_PREFER_PATH=' +
-                    self.spec['postgresql'].prefix.bin
+                    self.spec['postgresql'].prefix.bin,
+                    '-DPROJ_INCLUDE_DIR=' + self.spec['proj'].prefix.include,
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/qgis/package.py]: I usually use:
```suggestion
                    '-DPROJ_INCLUDE_DIR=' + self.spec['proj'].headers.directories[0]
```
but that might not be strictly necessary.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/qgis/package.py [hunk 3]
```diff
@@ -161,7 +161,10 @@ def cmake_args(self):
                     self.spec['geos'].prefix.bin,
                     '-DGSL_CONFIG_PREFER_PATH=' + self.spec['gsl'].prefix.bin,
                     '-DPOSTGRES_CONFIG_PREFER_PATH=' +
-                    self.spec['postgresql'].prefix.bin
+                    self.spec['postgresql'].prefix.bin,
+                    '-DPROJ_INCLUDE_DIR=' + self.spec['proj'].headers.directories[0]
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/qgis/package.py]:
```suggestion
                    '-DPROJ_INCLUDE_DIR=' + self.spec['proj'].headers.directories[0],
```

File: var/spack/repos/builtin/packages/qgis/package.py (hunk 1 above)
Comment (Sinan81) [reply var/spack/repos/builtin/packages/qgis/package.py]: great!

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**